### PR TITLE
Adding support for ensemble output format

### PR DESCRIPTION
**Pull request by anirban-chaudhuri**

### CODE REVIEW

**Review by sabinala — APPROVED** [review]
Nice!

**Comment by sabinala** [comment]
Why are all of the checks failing...oh no! We should fix that before merging.

**Comment by sabinala** [comment]
Especially the linting? When I enter `make format` it says all of the files are fine.

**Comment by SamWitty** [comment]
> Especially the linting? When I enter `make format` it says all of the files are fine.

@sabinala , `make format` does not automatically resolve typing inconsistencies, which is a good thing. Use `make lint` locally to see if the linter would pass.

**Comment by SamWitty** [comment]
@anirban-chaudhuri , I believe I resolved the current issues causing failing test in #572 . I created a separate PR because that contribution was independent of this one, although it is blocking.

File: pyciemss/integration_utils/result_processing.py [hunk 1]
```diff
@@ -86,9 +127,11 @@ def convert_to_output_format(
     }
 
     if timepoints is not None:
-        timepoints = [*timepoints]
+        # timepoints = [*timepoints]
```
Comment (SamWitty) [inline pyciemss/integration_utils/result_processing.py:130]: Please remove this commented out line

File: pyciemss/integration_utils/result_processing.py [hunk 2]
```diff
@@ -116,7 +159,190 @@ def convert_to_output_format(
     ):
         result = set_intervention_values(result, name, values, intervention_times)
 
-    return result
+    if ensemble_quantiles:
+        result_q = make_quantiles(
```
Comment (SamWitty) [inline pyciemss/integration_utils/result_processing.py:163]: Could you rename `result_q` to something a bit more informative?

File: pyciemss/integration_utils/result_processing.py [hunk 3]
```diff
@@ -116,7 +159,190 @@ def convert_to_output_format(
...
+                q["type"].extend(["quantile"] * num_timepoints * num_quantiles)
+                q["quantile"].extend(list(np.tile(alpha_qs, num_timepoints)))
+                q["value"].extend(
+                    list(
+                        np.squeeze(
+                            q_vals.T.reshape((num_timepoints * num_quantiles, 1))
+                        )
+                    )
+                )
+                # if "cum" in k.lower():
```
Comment (SamWitty) [inline pyciemss/integration_utils/result_processing.py:210]: Please remove commented out code

File: pyciemss/integration_utils/result_processing.py [hunk 4]
```diff
@@ -116,7 +159,190 @@ def convert_to_output_format(
...
+                )
+                q["output"].extend([k] * num_timepoints * num_quantiles)
+                q["type"].extend(["quantile"] * num_timepoints * num_quantiles)
+                q["quantile"].extend(list(np.repeat(alpha_qs, num_timepoints)))
+                q["value"].extend(
+                    list(
+                        np.squeeze(q_vals.reshape((num_timepoints * num_quantiles, 1)))
+                    )
+                )
+                # if "cum" in k.lower():
```
Comment (SamWitty) [inline pyciemss/integration_utils/result_processing.py:227]: Please remove commented out code

File: pyciemss/integration_utils/result_processing.py [hunk 5]
```diff
@@ -116,7 +159,190 @@ def convert_to_output_format(
...
+        if timepoints is not None:
+            all_timepoints = result_q["timepoint_id"].map(
+                lambda v: timepoints[v].item()
+            )
+            result_q = result_q.assign(**{f"number_{time_unit}": all_timepoints})
+            result_q = result_q[
+                [
+                    "timepoint_id",
+                    f"number_{time_unit}",
+                    # "inc_cum",
```
Comment (SamWitty) [inline pyciemss/integration_utils/result_processing.py:244]: please remove commented out code

File: pyciemss/interfaces.py [hunk 6]
```diff
@@ -52,6 +52,32 @@ def ensemble_sample(
     start_time: float = 0.0,
     inferred_parameters: Optional[pyro.nn.PyroModule] = None,
     time_unit: Optional[str] = None,
+    alpha_qs: Optional[List[float]] = [
```
Comment (SamWitty) [inline pyciemss/interfaces.py:55]: See previous comment about `DEFAULT_ALPHA_QS`.

File: pyciemss/interfaces.py [hunk 7]
```diff
@@ -151,7 +182,12 @@ def wrapped_model():
         )()
 
         return prepare_interchange_dictionary(
```
Comment (SamWitty) [inline pyciemss/interfaces.py:184]: I would strongly prefer for us to not include this very specific output processing in the definition of `sample`. Instead, let's have the ensemble_quantiles be generated separately if needed from the existing output of `sample`.

Comment (SamWitty) [reply pyciemss/interfaces.py:184]: This way we can mix and match output formats if they change in the future.

**Review by SamWitty — CHANGES_REQUESTED** [review]
This looks good overall. I've included some specific comments throughout.

My main general comment is that this specialized output processing should not be applied inside of the PyCIEMSS `sample` interfaces, as generating the samples and processing them are really separate computations. Instead let's just leave the new processing utilities as standalone functions, and apply them to the output of `sample` only when necessary.

File: pyciemss/integration_utils/result_processing.py [hunk 8]
```diff
@@ -1,25 +1,63 @@
...
+
 
 def prepare_interchange_dictionary(
     samples: Dict[str, torch.Tensor],
     time_unit: Optional[str] = None,
-    timepoints: Optional[Iterable[float]] = None,
+    timepoints: Optional[torch.Tensor] = None,
     visual_options: Union[None, bool, Dict[str, Any]] = None,
+    ensemble_quantiles: bool = False,
+    alpha_qs: Optional[List[float]] = [
```
[the remark this answers is not in the document]
Comment (anirban-chaudhuri) [reply pyciemss/integration_utils/result_processing.py:20]: done

File: pyciemss/interfaces.py [hunk 9]
```diff
@@ -33,6 +33,32 @@
 
 warnings.simplefilter("always", UserWarning)
 
+DEFAULT_ALPHA_QS = [
```
Comment (SamWitty) [inline pyciemss/interfaces.py:36]: Could you import this from `pyciemss.integration_utils.result_processing` instead of redefining it here?

**Review by SamWitty — CHANGES_REQUESTED** [review]
One minor change requested.

**Review by SamWitty — APPROVED** [review]

**Pull request merged**